src/python/espressomd/zn.py

```
@@ -0,0 +1,436 @@
+
+        # Catch when the server is initializing an empty frame
+        if self.frame_count == 0 and len(self.zndraw) > 0:
+            self.zndraw.__setitem__(0, znjson.dumps(
+                self.system, cls=znjson.ZnEncoder.from_converters(
+                    [EspressoConverter])
+            ))
+        else:
+            self.zndraw.append(znjson.dumps(
+                self.system, cls=znjson.ZnEncoder.from_converters(
+                    [EspressoConverter])
+            ))
+
+        if self.params["vector_field"] is not None and self.frame_count == 0:
+            self.zndraw.socket.sleep(1)
+            self.zndraw.config.arrows.colormap = self.arrow_config['colormap']
+            self.zndraw.config.arrows.normalize = self.arrow_config['normalize']
+            self.zndraw.config.arrows.colorrange = self.arrow_config['colorrange']
+            self.zndraw.config.arrows.scale_vector_thickness = self.arrow_config[
+                'scale_vector_thickness']
+            self.zndraw.config.arrows.opacity = self.arrow_config['opacity']
+
+        self.frame_count += 1
+
+    def draw_constraints(self, shapes: list):
+        """
+        Draw constraints on the visualizer
+        """
+        if not isinstance(shapes, list):
+            raise ValueError("Constraints must be given in a list")
+
+        objects = []
+
+        for shape in shapes:
+            if not isinstance(shape, t.Union[espressomd.shapes.Cylinder,
+                                             espressomd.shapes.Wall,
+                                             espressomd.shapes.Sphere,
+                                             espressomd.shapes.Rhomboid,
+                                             espressomd.shapes.Ellipsoid]):
+                raise ValueError(
+                    "Constraint must be of type Cylinder, Wall, Sphere, Rhomboid or Ellipsoid")
+
+            shape_type = shape.name()
+
+            if shape_type == "Shapes::Cylinder":
```

I'm not sure we can guarantee the stability of these internal names. For example, the `Caliper` class has name `"ScriptInterface::Profiler::Caliper"` (extra `ScriptInterface::` namespace). These namespaces may change in the future, and since ZnDraw doesn't run in CI, any change to internal names will go unnoticed and get merged into the main branch. Because this if-else chain has no final `else` (the equivalent of `default` in a switch statement), any shape whose internal name has changed will be silently skipped.

Wouldn't it be more sustainable to adapt these if-else to check against the Python name, i.e. `shape.__class__.__name__`, rather than the internal name, and have a final `else` that throws the `NotImplementedError` mentioned above? Then you don't need to maintain a type union at lines 281-285 anymore.